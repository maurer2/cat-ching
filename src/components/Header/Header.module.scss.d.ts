import globalClassNames, { ClassNames as GlobalClassNames } from '../..style.d';
declare const classNames: typeof globalClassNames & {
  readonly header: 'header';
  readonly title: 'title';
  readonly detail: 'detail';
  readonly detailKey: 'detailKey';
  readonly detailValue: 'detailValue';
  readonly resetButton: 'resetButton';
};
export default classNames;
export type ClassNames =
  | 'header'
  | 'title'
  | 'detail'
  | 'detailKey'
  | 'detailValue'
  | 'resetButton'
  | GlobalClassNames;
