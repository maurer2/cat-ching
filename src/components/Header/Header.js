import React from 'react';
import PropTypes from 'prop-types';

import style from './Header.module.scss';

function Header({ targetAmount, currentAmount, handleReset }) {
  return (
    <header className={style.header}>
      <h1 className={style.title}>
        Header
      </h1>
      <div className={style.details}>
        <dl className={style.detail}>
          <dt className={style.detailKey}>
            Target amount:
          </dt>
          <dd className={style.detailValue}>
            {targetAmount.valueFormatted}
          </dd>
        </dl>
        <dl className={style.detail}>
          <dt className={style.detailKey}>
            Current amount:
          </dt>
          <dd className={style.detailValue}>
            {currentAmount.valueFormatted}
          </dd>
        </dl>
      </div>
      <button
        className={style.resetButton}
        onClick={handleReset}
        type="button"
      >
        Reset
      </button>
    </header>
  );
}

export default Header;

const { string, number, func, shape } = PropTypes;

Header.propTypes = {
  targetAmount: shape({
    value: number.isRequired,
    currency: string.isRequired,
    valueInCents: number.isRequired,
    valueFormatted: string.isRequired,
  }).isRequired,
  currentAmount: shape({
    value: number.isRequired,
    currency: string.isRequired,
    valueInCents: number.isRequired,
    valueFormatted: string.isRequired,
  }).isRequired,
  handleReset: func.isRequired,
};
