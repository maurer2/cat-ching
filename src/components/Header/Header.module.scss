.header {
  display: flex;
  padding: 1rem;
  flex-wrap: wrap;
  align-items: center;
  background: black;
  color: white;
}
  
.title {
  margin-top: 0;
  flex-basis: 100%;
  text-align: center;
  font-size: 1.5rem;
}

.details {
  flex-grow: 1;
}

.detail {
  display: flex;
  margin: 0;
  flex-wrap: wrap;

  & + .detail {
    margin-top: 0.5rem;
  }
}

.detailKey {
  margin-right: 0.5rem;
  flex-basis: auto;
}

.detailValue {
  margin: 0;
  font-weight: bold;
}

.resetButton {
  padding: 0.5rem;
  background: white;
  color: black;
}
