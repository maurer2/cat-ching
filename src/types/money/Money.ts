/* eslint-disable no-underscore-dangle */
/* eslint-disable @typescript-eslint/lines-between-class-members */
export const currencies = {
  GBP: 'Â£',
  USD: '$',
} as const;

export type CurrencyName = keyof typeof currencies;

export default class Money {
  private readonly _value: bigint;
  private readonly _name: CurrencyName;
  private readonly _formatter: Intl.NumberFormat;

  public static fromBigInt(
    valueInFractions: bigint,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    return new this(valueInFractions, name, formatter);
  }

  public static fromNumber(
    valueInFractions: number,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    const valueAsBigInt = BigInt(valueInFractions);

    return new this(valueAsBigInt, name, formatter);
  }

  public static fromRandom(name: CurrencyName, formatter?: Intl.NumberFormat): Money {
    const randomNumber: number = Math.ceil(Math.random() * (1000 - 1)) + 1;
    const valueAsBigInt = BigInt(randomNumber);

    return new this(valueAsBigInt, name, formatter);
  }

  private constructor(valueInFractions: bigint, name: CurrencyName, formatter?: Intl.NumberFormat) {
    this._value = valueInFractions;
    this._name = name;
    this._formatter = formatter
      ?? new Intl.NumberFormat('en-GB', {
        style: 'currency',
        currency: 'GBP',
        useGrouping: false, // disable thousand separator
      });
  }

  public add(valueToAdd: Money): Money {
    const newValue: bigint = this._value + valueToAdd._value;

    return new Money(newValue, this._name, this._formatter);
  }

  public isEqualTo(secondValue: Money): boolean {
    return this._value === secondValue._value;
  }

  public isNegative(): boolean {
    return this._value < 0;
  }

  public get valueAsSeparateParts(): Intl.NumberFormatPart[] {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.formatToParts(valueAsNumber);
  }

  public get valueAsFormattedString(): string {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.format(valueAsNumber);
  }

  public get valueFormattedAsIntegerAndFraction(): string {
    // https://stackoverflow.com/a/70086020
    const integer = Number(this._value / 100n);
    const fraction = Number(this._value % 100n);
    // const integral = Number(this.value / 100n);
    // const fraction = Number(this.value % 100n);

    return `${integer}.${`${fraction}`.padEnd(2, '0')}`;
  }
}

export type MoneyType = typeof Money;
