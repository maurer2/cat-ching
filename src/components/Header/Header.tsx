import React from 'react';

import Money from '../../types/Money';

import style from './Header.module.scss';

function Header({ targetAmount, handleReset }) {
  const targetValue = (targetAmount as Money).valueAsFormattedString;

  return (
    <header className={style.header}>
      <h1 className={style.title}>Cat-Ching</h1>
      <div className={style.details}>
        <dl className={style.detail}>
          <dt className={style.detailKey}>Target amount:</dt>
          <dd className={style.detailValue}>{targetValue}</dd>
        </dl>
      </div>
      <button
        className={style.resetButton}
        onClick={handleReset}
        type="button"
      >
        Reset
      </button>
    </header>
  );
}

export default Header;
