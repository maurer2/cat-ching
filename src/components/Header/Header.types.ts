import Money from '../../types/Money';

export type HeaderProps = {
  targetAmount: Money;
  handleReset: () => void;
};
