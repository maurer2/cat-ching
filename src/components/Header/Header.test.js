import React from 'react';
import { shallow } from 'enzyme';

import Header from './Header';
import Money from '../../data/money';

describe('Header', () => {
  const moneyData = new Money(1, 'Dollar');

  const mockedHandleReset = jest.fn();

  it('should render correctly', () => {
    const wrapper = shallow(<Header
      targetAmount={moneyData}
      handleReset={mockedHandleReset}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
