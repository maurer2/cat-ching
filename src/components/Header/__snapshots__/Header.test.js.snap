// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<header
  className="header"
>
  <h1
    className="title"
  >
    Cat-Ching
  </h1>
  <div
    className="details"
  >
    <dl
      className="detail"
    >
      <dt
        className="detailKey"
      >
        Target amount:
      </dt>
      <dd
        className="detailValue"
      >
        0.01
      </dd>
    </dl>
  </div>
  <button
    className="resetButton"
    onClick={[MockFunction]}
    type="button"
  >
    Reset
  </button>
</header>
`;
